### ISSUE TITLE
Swarm HA may not work properly with Mesos

### ISSUE THREAD

**Issue by ezrasilvera** [docker/swarm#2000]
In manage.go all managers (leader and followers) first create the cluster (mesos.NewCluster() )  and only then decide if their role is leader or follower (through setupReplication()  ).
In mesos.NewCluster() we register the Swarm framework in Mesos.

As a result of the above it seem that also the "followers" will register in Mesos and may mixup the Mesos scheduler (e.g., recive offeres, tec..)

**Comment by Spritekin** [comment]
I was observing at this as well. This is not a "may not work", but a "does not work". It is exactly as you say.

I thought it was a side effect of my changes here https://github.com/docker/swarm/pull/2067 but it looks like this code has to be improved.

One of the problems is that swarm was not specifying a frameworkid so mesos would assign a random ID. As an effect if I have 5 swarm services each is considered a different framework. 
The PR https://github.com/docker/swarm/pull/2067, adds the capacity to specify the ID so any number of swarm instances are considered the same framework.

The second problem is that swarm was not specifying a task timeout. This means that if swarm crashed, all the jobs started from that instance were immediately terminated. By providing "generous" time for persistence (as defined in the mesos framework guidelines which recommend many days) swarm can crash and the tasks will keep running. If the framework is restarted with the same framework id as before, mesos will reattach the tasks.

With those two pieces in place, this one makes sense. The procedure would be:
a. Start multiple swarm instances following the HA guidelines (ZK/consul for coordination)
b. Each instance should define the same FrameworkID so mesos considers them the same.
c. Instances should NOT connect to mesos but live in a pending state.
d. If the leader dies, it must communicate with the slaves so one takes the lead and connects to mesos.

a. One thing to observe is that the swarm service is running from marathon then health checks are a MUST. They should be very aggressive. I have had times where the /info http endpoint responds but the /run doesn't work not even running hello-world. In this case the leader should be terminated.

b. As a result of (a) the question is, what is faster, to allow marathon to restart the service or to allow reelection. I think the point is not how fast the service restarts but how fast can we detect a swarm failure. If marathon has to wait for 3 checks before shutting down a service then it may be 3 mins with no service, plus the time it takes to restart plus the time it takes for the DNS to refresh.

c. If the sleeping services are acting as proxies and redirect requests to the leader, then the bottleneck is the leader.

d. I think the best solution would be running 5 swarm services in parallel, each one with its own frameworkid. If any fails, marathon can restart it with the same framework id so it can be restarted without task loss. An nginx can be set as load balancer jumping between services.
The problem with this approach is that swarm is resource offer greedy. It captures an offer and won't let go for many seconds. If there are 5 services with the same offer, one gets the offer for 15 seconds and it won't let go. If there are 5 swarm nodes one will get the offer after 1 minute which is VERY slow. Offers should be captured at most for 1 second. This leads to new problems I don't want to talk about.

So in other words, I decided to give HA the boot and I only have 1 swarm master started by marathon with a specific framework ID, if it crashes marathon can restart with the same id, jobs have a wait period of 3 days so hey will survive while swarm reconnects and everybody happy. Marathon health checks run every 30secs and accept 2 failures and my DNS has a refresh rate 10 secs, so if the master fails it will restart within 1 min10s which is fine for me.

**Comment by ezrasilvera** [comment]
So just to clarify - for now you just added the framework-id. What happen if I have several frameworks registered with the same ID ? Is that possible in Mesos or I'll get an error ?

**Comment by Spritekin** [comment]
I have started multiple frameworks with the same ID. They all appear in the task tab, but only one appears in the Frameworks tab. Its like mesos has them unified under the same ID.

The problem I had now is exactly what you mentioned. When a secondary swarm registers, mesos will send the offers to that secondary. What I guess happens next is that the leader receives the job requests but it never gets the offers because they are captured by the secondary services.

**Comment by ezrasilvera** [comment]
@Spritekin I have started to look into a possible fix for that .. will try to push a PR in the comming few days ..

**Comment by nishanttotla** [comment]
Closing due to no activity and https://github.com/docker/swarm/pull/2853
